Drop-In/src/main/java/com/braintreepayments/api/DropInResult2.java

```diff
@@ -0,0 +1,61 @@
+package com.braintreepayments.api;
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import androidx.annotation.Nullable;
+
+public class DropInResult2 implements Parcelable  {
```

Comment:
Looking for input here. The activity contract requires a single result object. We could add error as a field on the existing `DropInResult` object, but that could cause confusion for existing integrations not using this Activity Result API. I decided to create a wrapper result that contains a `DropInResult` or `Exception` to keep it simple and separate the integrations. Couldn't think of a good name for this new result so looking for suggestions on that as well.

Reply:
I think this is a good call. `DropInActivityResult` might be an option as well, though it's really similar to `DropInResult`. If this is the recommended route though maybe we can mark the previous result class a deprecated? Not sure how that would work or if it makes sense but we have a few options.

Reply:
Hmm.. I marked it as deprecated, but am also wondering if that will cause confusion for merchants unless we update our docs and migration guide to essentially ignore the `DropInClient#launchDropInForResult` option. It might be a big lift to ask merchants to upgrade to v6 and adopt the new Activity Result APIs. Maybe we should leave them both as options for now and not deprecate?

Reply:
Ah yeah that's a good point. Maybe we could keep `DropInResult` then and add the `error` field? We can still keep the error `Intent` extra while also adding it to the `DropInResult` parcelable. A bit redundant, but it would prevent a breaking change.

Reply:
I also like the idea of creating this wrapper class and not muddying up the existing `DropInResult`. 

Another option is that we mark every component of the old flow (non activity result API) as deprecated, and then we do include these updates into the existing `DropInResult`. 

Also I like Steven's naming suggestion of `DropInActivityResult` if we are going forward with a separate class. 

Either way, I think we should definitely deprecate the old flow and update our docs at the same time. I think having 2 flows that do the same thing is confusing. And deprecating doesn't cause anything to break, just signals to merchants "hey you should update this sooner than later".

Reply:
Ok I am going to go with the wrapper class and rename it to `DropInActivityResult`, and deprecate `DropInClient#launchDropInForResult` 👍

Reply:
Shouldn't we also deprecate the entire `DropInResult` too?

Reply:
I don't think it should be deprecated since `DropInActivityResult` contains a `DropInResult` or an `Exception` - that `DropInResult` still holds the actual payment method type, nonce, etc. In the deprecated flow the `DropInResult` would be accessed from an intent extra in `onActivityResult`, and now it will be accessed from the `DropInActivityResult` produced by the `DropInActivityContract`. Do you think that's clear enough?

Reply:
Let me know if that idea makes sense - but basically it's in favor of duplicating now, so that the deprecation / update path is more clear.

Reply:
Hmm I still don't think I understand the need to deprecate `DropInResult` at all, it's not really impacted by this change. The only reason this `DropInActivityResult` class is added is because this approach requires a single return object, and it seemed cleaner for the merchant code to have the single object either have a non-null result, or a non-null error. We could alternatively just add the error as a parameter on the original `DropInResult`, but then the null-checking in the merchant code isn't quite as simple. It's not a huge difference but I think it would be something like this:

```
        resultLauncher = registerForActivityResult(DropInActivityContract()) { result ->
            result?.let {
                it.dropInResult?.let { dropInResult ->
                    // handle result accessing dropInResult.paymentMethodNonce?, dropInResult.paymentMethodType?, etc
                }
                it.error?.let { dropInError ->
                    // handle error
                }
            }
            // user cancelled drop in
        }
```
vs: 
```
        resultLauncher = registerForActivityResult(DropInActivityContract()) { result ->
            result?.let {
                it.paymentMethodNonce?.let { paymentMethodNonce ->
                    // handle result using fields on outer-scoped it (it.paymentMethodType?, it.paymentMethodDescription?) and paymentMethodNonce
                }
                it.error?.let { dropInError ->
                    // handle error
                }
            }
            // user cancelled drop in
        }
```

Maybe the difference isn't a big enough reason to add this wrapper type. What do you think?

Reply:
@sshropshire Sammy and I discussed this a bit more and decided that the single `DropInActivityResult` object makes sense, and we will keep the `DropInResult` not deprecated since it contains the result information. We are now considering the cancelation handling with this new flow. Previously, in `onActivityResult`, the merchant would received an explicit `RESULT_CANCELED` result code so that they could handle the cancellation case. In this flow, we are kind of lumping cancellations into any null result, and aren't explicitly telling a merchant that the result was cancelled.  There are a few ways we could handle this:
1) Keep it as is, and allow merchants to handle cancellation along with any other null result outcomes (but give no explicit cancellation information)
2) Add an `isCanceled` boolean to the `DropInActivityResult` 
3) Return cancelation as a `UserCanceledException` in the `error` field already on `DropInActivityResult`

Do you have an opinion on which approach would be best?

Reply:
I think any option could work. If possible I'd wrap the androidx `registerForActivityResult` method to give us ownership of the callback method signature. Then we could return`(DropInResult, Error)` to the merchant without them knowing we're using the register activity API under the hood.

Reply:
Since `registerForActivityResult` is a method on `ComponentActivity` and `Fragment` I'm not sure that we could do that. We couldn't allow them to use DropInClient to call it at least. We'd have to create our own Fragment and Activity that merchants would override I think.

Reply:
Yeah that's true. We would have to have multiple overloads for the start method. Like `launchDropInForResult(ComponentActivity)` and `launchDropInForResult(Fragment)`.

Reply:
Ah good idea! I spec'd out what that would look like with a `DropInLauncher` class. It's a single commit (without updating tests) so can revert if we prefer giving merchants full control over the Activity Result API.
Pros to `DropInLauncher`:
- We can keep the `DropInActivityResult` class internal and pass back a tuple (result, error) to a callback similar to all other APIs
- We can make `DropInActivityContract` non-public and hide the actual Activity Result API logic from the merchant

Cons to `DropInLauncher`:
- Merchants need to create both a `DropInClient` and a `DropInLauncher` to launch Drop-in (`DropInClient` can't easily be remove/instantiated within `DropInLauncher` because we need to call `DropInLauncher#registerForActivityResult` before the fragment/activity is created, but we need to give merchants a chance to fetch authorization before creating a `DropInClient`
- We take responsibility for documenting when to call the lifecycle methods (i.e. call `DropInLauncher#registerForActivityResult` in `onCreate` and make sure to call it before you call `DropInLauncher.launchWithDropInClient`) instead of letting Google Activity Result API document the requirements. 

Let me know what you think/if I'm missing anything else with this approach.

Reply:
This is super helpful, thanks for outlining it so clearly. I guess we need to decide which is more important - more granular/specific return type for merchants or more control over activity launch lifecycle.